### PR TITLE
Standardize and normalize packages

### LINKED ISSUE #17: Normalize and standardise manifests
pacote should normalize metadata fields to only the things the CLI might need, and standardize those across the different sources. Should also run the manifest through `normalize-package-data`.

### PR DESCRIPTION
**Pull request by zkat**
Fixes #17 and #18.

With this PR, `pacote.manifest` will now return a single, immutable `Manifest` object that has a consistent set of fields, with normalized data and semantics. These fields are chosen based on their utility to the installer. Things that are missing are thus not considered necessary for installation.

Aside from a few handler-specific details, there's now a single place in the code where we do fallback to the package tarballs in order to fill in missing/unknown fields, including `_shrinkwrap`, `bin`, and even `_shasum`.

### TODO

- [x] write dedicated tests for `finalize-manifest`. Include tests to check the bypassing of tarball extracts
- [x] convert `extract-shrinkwrap` tests to be used by `finalize-manifest` tests.

### CODE REVIEW

File: lib/registry/manifest.js [hunk 1]
```diff
@@ -28,13 +26,21 @@ function manifest (spec, opts, cb) {
...
-        'registry.manifest',
-        spec.name + '@' + spec.spec,
-        'resolved to',
-        manifest.name + '@' + manifest.version
+      // Done here instead of ./finalize-manifest because these fields
+      // have to be filled in differently depending on type.
+      manifest._shasum = (
+        manifest.dist && manifest.dist.shasum
+      ) || manifest._shasum
+      manifest._resolved = (
```
Comment (iarna) [inline lib/registry/manifest.js:34]: Your lisp accent is showing. 😉   I feel like these expressions are waaay too big to be easy to follow in JS. If you don't want to do if-branched assignment then maybe put the different alternatives in functions?

File: lib/finalize-manifest.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    pkg.bundleDependencies ||
+    false
+  )
+  this.bundleDependencies = bundled
+  this.peerDependencies = pkg.peerDependencies || {}
+
+  // This one depends entirely on each handler.
+  this._resolved = pkg._resolved
+
+  // We sometimes need to grab these from the tarball.
```
Comment (varjmes) [inline lib/finalize-manifest.js]: Why?

File: lib/finalize-manifest.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  this._shasum = pkg._shasum || fromTarball._shasum
+  this._shrinkwrap = pkg._shrinkwrap || fromTarball._shrinkwrap || null
+  this.bin = pkg.bin || fromTarball.bin || null
+
+  if (this.bin && Array.isArray(this.bin)) {
+    // This isn't even documented?
```
Comment (varjmes) [inline lib/finalize-manifest.js]: Is this comment necessary?

File: lib/finalize-manifest.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  this._id = null // filled in by normalize-package-data, but unnecessary
+
+  // TODO - freezing and inextensibility pending npm changes. See test suite.
+  // Object.preventExtensions(this)
+  normalize(this)
+
+  // I don't want this why did you give it to me. Go away. 🔥🔥🔥🔥
+  delete this.readme
+
+  // Object.freeze(this)
```
Comment (varjmes) [inline lib/finalize-manifest.js:73]: Do you need this comment? :)

File: test/extract-stream.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+require('./util/test-dir')(__filename)
+
+var extractStream = require('../lib/extract-stream')
+
+npmlog.level = process.env.LOGLEVEL || 'silent'
+var OPTS = {
+  log: npmlog
+}
+
+test('basic extraction')
```
Comment (varjmes) [inline test/extract-stream.js:17]: If it hasn't already, let's create an issue to finish these tests lest we forget.

File: lib/finalize-manifest.js (hunk 4 above)
Comment (zkat) [reply lib/finalize-manifest.js:73]: This is a reminder to myself of something I really want to do but npm doesn't let me do yet -- it'll be incommented once I figure a few things out, and I didn't want it to be forgotten.

File: lib/finalize-manifest.js (hunk 2 above)
Comment (zkat) [reply lib/finalize-manifest.js]: I'll expand on this. I think it's worth explaining.

File: lib/finalize-manifest.js (hunk 3 above)
Comment (zkat) [reply lib/finalize-manifest.js]: No. I was just sad.

File: test/extract-stream.js (hunk 5 above)
Comment (zkat) [reply test/extract-stream.js:17]: Not gonna forget. These are *very visible* if you run `npm t`. Also, #48 is the PR adding these tests (and `registry.tarball.js`)

File: lib/finalize-manifest.js (hunk 4 above)
Comment (varjmes) [reply lib/finalize-manifest.js:73]: Ok, thanks :)

File: test/extract-stream.js (hunk 5 above)
Comment (varjmes) [reply test/extract-stream.js:17]: My bad. Thanks!

File: lib/finalize-manifest.js (hunk 3 above)
Comment (varjmes) [reply lib/finalize-manifest.js]: :( <3

**Review by varjmes — APPROVED** [review]

**Comment by iarna** [comment]
The bin thing totally _is_ documented. See the `directories` section of the `package.json` docs:

> If you specify a `bin` directory in `directories.bin`, all the files in that folder will be added.
> 
> Because of the way the `bin` directive works, specifying both a `bin` path and setting `directories.bin` is an error. If you want to specify individual files, use `bin`, and for all the files in an existing `bin` directory, use `directories.bin`.

**Review by iarna — APPROVED** [review]

**Pull request merged**